core/src/main/java/org/apache/accumulo/fate/zookeeper/DistributedReadWriteLock.java

```diff
@@ -218,22 +233,82 @@ public boolean tryLock() {
       }
       SortedMap<Long,byte[]> entries = qlock.getEarlierEntries(entry);
       Iterator<Entry<Long,byte[]>> iterator = entries.entrySet().iterator();
-      if (!iterator.hasNext())
+      if (!iterator.hasNext()) {
         throw new IllegalStateException("Did not find our own lock in the queue: " + this.entry
             + " userData " + new String(this.userData, UTF_8) + " lockType " + lockType());
-      return iterator.next().getKey().equals(entry);
+      }
+      if (!failBlockers) {
+        return iterator.next().getKey().equals(entry);
+      } else {
+        ZooStore<DistributedReadWriteLock> zs;
+        try {
+          zs = new ZooStore<>(zooPath, zrw);
+        } catch (KeeperException | InterruptedException e1) {
+          log.error("Error creating zoo store", e1);
+          return false;
+        }
+        // Loop through all of the prior transactions that are waiting to
+        // acquire this write lock. If the transaction has not succeeded or failed,
+        // then fail it and return false from this method so that Utils.reserveX()
+        // will call this method again and will re-check the prior transactions.
+        boolean result = true;
+        while (iterator.hasNext()) {
+          Entry<Long,byte[]> e = iterator.next();
+          Long txid = e.getKey();
+          if (!txid.equals(entry)) {
+            if (zs.isReserved(txid)) {
+              result &= false;
+            } else {
+              zs.reserve(txid);
+              TStatus status = zs.getStatus(txid);
+              if (status.equals(TStatus.FAILED)) {
+                result = false;
+                continue;
+              } else {
+                switch (status) {
+                  case UNKNOWN:
+                    // not sure what to do here, we have an invalid txid
+                    break;
+                  case SUCCESSFUL:
+                    // already succeeded, lock should be removed
+                    break;
+                  case FAILED:
+                  case FAILED_IN_PROGRESS:
```

Comment:
Investigation would be needed to confirm this, but I suspect we should only expect to see IN_PROGRESS and FAILED_IN_PROGRESS transactions actually holding the lock.  Anything else seems like and indication of a more general problem that we should not attempt to react to here.  We could log a warn/error if we see them, but it may not be a good idea to let the lock be acquired when the system is an unexpected state. This kinda goes w/ my other comment of always returning false when we are not first in the queue.

Reply:
Addressed in 2c3ea13